### PR TITLE
[5.2] Enable including and excluding articles in mod_articles

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #44150 .

### Summary of Changes
With this PR a user can include and exclude articles and exclude the current article

### Testing Instructions
Playing with all variants of included and excluded article Ids. For example exclude articles and iclude them at the same time and use all possible combinations of excluding and including articles. 
The result should be consistent. 

### Actual result BEFORE applying this Pull Request
A user can either include or exclude articles. He can exclude the current article only if the exclude variant is active, but then he cannot include anything else 

### Expected result AFTER applying this Pull Request
The user can exclude the current article.
He can exclude also other article Ids
He can include other articles
If the same article is included and excluded, this id is ignored

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 1]
```diff
@@ -212,36 +212,55 @@ public function getArticles(Registry $params, SiteApplication $app)
...
-            $articlesList[] = (int) $article['id'];
+        foreach (ArrayHelper::fromObject($params->get('excluded_articles', '')) as $article) {
+            $excludedList[] = (int) $article['id'];
         }
 
-        if (!empty($articlesList)) {
-            $articles->setState('filter.article_id', $articlesList);
-            $articles->setState('filter.article_id.include', $filterInclude);
+        if ($params->get('exclude_current', 1) && $activeArticle !== 0) {
+             // Add the id of the current article to the excluded list
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
            // Add the id of the current article to the excluded list
```
Fix PHPCS (1 of 4)

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 2]
```diff
@@ -212,36 +212,55 @@ public function getArticles(Registry $params, SiteApplication $app)
...
+        if (!empty($includedList) && !empty($excludedList)) {
+
+            // Remove duplicates from the list. If an article is excluded and included, ignore the id.
+            $tmp1 = array_diff($includedList, $excludedList); // Values only in $array1
+            $tmp2 = array_diff($excludedList, $includedList); // Values only in $array2
+
+            // Merge the unique values
+            $articlesList = array_merge($tmp1, $tmp2);
+            
+            // Separate the lists again so we can set filters
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
            $articlesList = array_merge($tmp1, $tmp2);

            // Separate the lists again so we can set filters
```
PHPCS (2 of 4)

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 3]
```diff
@@ -212,36 +212,55 @@ public function getArticles(Registry $params, SiteApplication $app)
...
+            foreach ($articlesList as $value) {
+                if (in_array($value, $includedList)) {
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
                if (\in_array($value, $includedList)) {
```
PHPCS (3 of 4)

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 4]
```diff
@@ -267,19 +286,17 @@ public function getArticles(Registry $params, SiteApplication $app)
...
-        $view   = $input->get('view');
-
-        if ($option === 'com_content' && $view === 'article') {
-            $active_article_id = $input->getInt('id');
-        } else {
-            $active_article_id = 0;
+            
+        // Eliminate excluded items
+        foreach ($items as $key => $item) {
+            if (in_array($item->id, $filterExclude)) {
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
            if (\in_array($item->id, $filterExclude)) {
```
PHPCS (4 of 4)

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 5]
```diff
@@ -267,19 +286,17 @@ public function getArticles(Registry $params, SiteApplication $app)
...
-        // Find current Article ID if on an article page
-        $option = $input->get('option');
-        $view   = $input->get('view');
-
-        if ($option === 'com_content' && $view === 'article') {
-            $active_article_id = $input->getInt('id');
-        } else {
-            $active_article_id = 0;
+            
+        // Eliminate excluded items
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion

        // Eliminate excluded items
```
PHPCS

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 6]
```diff
@@ -212,36 +212,55 @@ public function getArticles(Registry $params, SiteApplication $app)
...
+            // Add the id of the current article to the excluded list
+            $excludedList[] = $activeArticle;
         }
 
+        // Build the filter lists
+        $filterInclude = [];
+        $filterExclude = [];
+
+        if (!empty($includedList) && !empty($excludedList)) {
+
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
```
Now as the first PHPCS checker is happy, the other one complains. Fix 1 of 2.

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 7]
```diff
@@ -268,18 +287,16 @@ public function getArticles(Registry $params, SiteApplication $app)
...
+        // Eliminate excluded items
+        foreach ($items as $key => $item) {
+            if (\in_array($item->id, $filterExclude)) {
+                unset($items[$key]);
+            }
         }
 
         // Prepare data for display using display options
         foreach ($items as &$item) {
+
```
Comment (richard67) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
```
Now as the first PHPCS checker is happy, the other one complains. Fix 2 of 2.

File: modules/mod_articles/src/Helper/ArticlesHelper.php [hunk 8]
```diff
@@ -338,7 +353,7 @@ public function getArticles(Registry $params, SiteApplication $app)
             }
 
             // Used for styling the active article
-            $item->active      = $item->id == $active_article_id ? 'active' : '';
+            $item->active      = $item->id == $activeArticle ? 'active' : '';
```
Comment (Quy) [inline modules/mod_articles/src/Helper/ArticlesHelper.php]:
```suggestion
            $item->active = $item->id == $activeArticle ? 'active' : '';
```

File: modules/mod_articles/mod_articles.xml [hunk 9]
```diff
@@ -105,25 +105,13 @@
 					showon="show_child_category_articles:1"
 				/>
 
-				<field
-					name="ex_or_include_articles"
-					type="list"
-					label="MOD_ARTICLES_FIELD_EX_OR_INCLUDE_LABEL"
```
Comment (Quy) [inline modules/mod_articles/mod_articles.xml:111]: Deprecate string.

Comment (brianteeman) [reply modules/mod_articles/mod_articles.xml:111]: its a new string and can simply be deleted as it has never been in a release

**Comment by rdeutz** [comment]
please check this one: https://github.com/joomla/joomla-cms/pull/44213

**Pull request closed without merging**